File: Python/Programmars/Level_1/키패드 누르기.py
def move_left(i, j):
    global left, answer
    left = (i, j)
    answer += 'L'


def move_right(i, j):
    global right, answer
    right = (i, j)
    answer += 'R'


def solution(numbers, hand):
    global answer, keypad, left, right
    answer = ''
    keypad = [[1, 2, 3], [4, 5, 6], [7, 8, 9], [-1, 0, -1]]
    left = (3, 0)
    right = (3, 2)
    for num in numbers:
        for i in range(4):
            for j in range(3):
                if keypad[i][j] == num:
                    if j == 0:
                        move_left(i, j)
                    elif j == 2:
                        move_right(i, j)
                    else:
                        left_dis = abs(left[0]-i) + abs(left[1]-j)
                        right_dis = abs(right[0]-i) + abs(right[1]-j)
                        if left_dis > right_dis:
                            move_right(i, j)
                        elif left_dis < right_dis:
                            move_left(i, j)
                        else:
                            if hand == "right":
                                move_right(i, j)
                            else:
                                move_left(i, j)
    return answer


print(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"))  # "LRLLLRLLRRL"
print(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], "left"))  # "LRLLRRLLLRR"
print(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0], "right"))  # "LLRLLRLLRL"

File: JavaScript/Programmars/카카오인턴십/키패드 누르기.js
class Keypad {
  pad = {
    1: [0, 0],
    2: [0, 1],
    3: [0, 2],
    4: [1, 0],
    5: [1, 1],
    6: [1, 2],
    7: [2, 0],
    8: [2, 1],
    9: [2, 2],
    '*': [3, 0],
    0: [3, 1],
    '#': [3, 2],
  };

  position(n) {
    return this.pad[n];
  }

  distance([x1, y1], [x2, y2]) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
  }
}

class Thumbs {
  constructor(handed, keypad) {
    this.left = '*';
    this.right = '#';
    this.handed = handed;
    this.keypad = keypad;
  }

  move(n) {
    if (n === 1 || n === 4 || n === 7) {
      return this.push('left', n);
    } else if (n === 3 || n === 6 || n === 9) {
      return this.push('right', n);
    } else {
      return this.select(n);
    }
  }

  push(direction, n) {
    if (direction === 'left') {
      this.left = n;
    } else {
      this.right = n;
    }
    return direction === 'left' ? 'L' : 'R';
  }

  select(n) {
    const left = this.keypad.position(this.left);
    const right = this.keypad.position(this.right);
    const target = this.keypad.position(n);

    const leftDiff = this.keypad.distance(left, target);
    const rightDiff = this.keypad.distance(right, target);

    if (leftDiff < rightDiff) {
      return this.push('left', n);
    } else if (leftDiff > rightDiff) {
      return this.push('right', n);
    } else {
      return this.push(this.handed, n);
    }
  }
}

function solution(numbers, hand) {
  let answer = '';
  const keypad = new Keypad();
  const thumbs = new Thumbs(hand, keypad);

  for (let num of numbers) {
    answer += thumbs.move(num);
  }

  return answer;
}

console.log(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], 'right')); // "LRLLLRLLRRL"
console.log(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], 'left')); // "LRLLRRLLLRR"
console.log(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0], 'right')); // "LLRLLRLLRL"
